src/Workspaces/Core/Portable/EmbeddedLanguages/RegularExpressions/RegexParser.cs

```diff
@@ -0,0 +1,1927 @@
+            // [b-\-a] and [b-\-\-a].  That's because the range that is checked is actually "b-a", even
+            // though it has all the \- escapes in the middle.
+
+            var first = ParseSingleCharacterClassComponent(isFirst: false, afterRangeMinus: true);
+            if (!IsEscapedMinus(first))
+            {
+                return first;
+            }
+
+            var builder = ArrayBuilder<RegexExpressionNode>.GetInstance();
+            builder.Add(first);
+
+            while (IsEscapedMinus(builder.Last()) && _currentToken.Kind != RegexKind.CloseBracketToken)
+            {
+                builder.Add(ParseSingleCharacterClassComponent(isFirst: false, afterRangeMinus: true));
+            }
+
+            return new RegexSequenceNode(builder.ToImmutable());
+        }
+
+        private RegexPrimaryExpressionNode ParseSingleCharacterClassComponent(bool isFirst, bool afterRangeMinus)
+        {
+            if (_currentToken.Kind == RegexKind.BackslashToken && _lexer.Position < _lexer.Text.Length)
+            {
+                var backslashToken = _currentToken;
+                var afterSlash = _lexer.Position;
+
+                // trivia is not allowed anywhere in a character class, and definitely not between
+                // a \ and the following character.
+                ConsumeCurrentToken(allowTrivia: false);
+                Debug.Assert(_currentToken.VirtualChars.Length == 1);
+
+                var nextChar = _currentToken.VirtualChars[0].Char;
+                switch (nextChar)
+                {
+                    case 'D': case 'd':
+                    case 'S': case 's':
+                    case 'W': case 'w':
+                    case 'p': case 'P':
+                        if (afterRangeMinus)
+                        {
+                            backslashToken = backslashToken.AddDiagnosticIfNone(new EmbeddedDiagnostic(
+                                string.Format(WorkspacesResources.Cannot_include_class_0_in_character_range, nextChar),
+                                GetSpan(backslashToken, _currentToken)));
+                        }
+
+                        // move back before the character we just scanned.
+                        // trivia is not allowed anywhere in a character class
+                        _lexer.Position--;
+                        return ParseEscape(backslashToken, allowTriviaAfterEnd: false);
+
+                    case '-':
+                        // trivia is not allowed anywhere in a character class
+                        return new RegexSimpleEscapeNode(
+                            backslashToken, ConsumeCurrentToken(allowTrivia: false).With(kind: RegexKind.TextToken));
+
+                    default:
+                        // trivia is not allowed anywhere in a character class
+                        _lexer.Position--;
+                        return ScanCharEscape(backslashToken, allowTriviaAfterEnd: false);
```

Comment:
Why is `ScanCharEscape` called `Scan` (rather than `Parse`) and could the name be improved to clarify `ScanCharEscape` versus `ParseEscape`?

Reply:
1.  Naming.  Yes, i should just name these Parse.  'Scan' came from the native impl, but i'm not beholden to that.
2. I will add comment explaining what's going on here, and how it is intentional.  Specifically, inside a character-class, you can only have Character-Escapes (not all escapes).  Also, the escapes are interpreted differently.  For example, outside of a character class \b is an anchor.  Inside a character-class \b is a simple-escape.  Similarly, \k'...'  is a k-capture escape outside of a char-class.  Inside a char-class it doesn't have special meaning.

Will add hte right docs.

Reply:
#resolved